### PR TITLE
DOC: update the dtypes/ftypes docstring (Seoul)

### PR DESCRIPTION
**Pull request by jongwony**
Checklist for the pandas documentation sprint (ignore this if you are doing
an unrelated PR):

- [X] PR title is "DOC: update the <your-function-or-method> docstring"
- [X] The validation script passes: `scripts/validate_docstrings.py <your-function-or-method>`
- [X] The PEP8 style check passes: `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] The html version looks good: `python doc/make.py --single <your-function-or-method>`
- [X] It has been proofread on language by another sprint participant

Please include the output of the validation script below between the "```" ticks:

```
# paste output of "scripts/validate_docstrings.py <your-function-or-method>" here
# between the "```" (remove this comment, but keep the "```")


################################################################################
##################### Docstring (pandas.DataFrame.dtypes)  #####################
################################################################################

Return the dtypes in this object.

Notes
-----
It returns a Series with the data type of each column.
If object contains data multiple dtypes in a single column,
dtypes will be chosen to accommodate all of the data types.
``object`` is the most general.

Examples
--------
>>> df = pd.DataFrame({'f': pd.np.random.rand(3),
...                    'i': 1,
...                    'd': pd.Timestamp('20180310'),
...                    'o': 'foo'})
>>> df.dtypes
f           float64
i             int64
d    datetime64[ns]
o            object
dtype: object

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
	No extended summary found
	No returns section found
	See Also section not found


################################################################################
##################### Docstring (pandas.DataFrame.ftypes)  #####################
################################################################################

Return the ftypes (indication of sparse/dense and dtype)
in this object.

Notes
-----
Sparse data should have the same dtypes as its dense representation

See Also
--------
dtypes, SparseDataFrame

Examples
--------
>>> arr = pd.np.random.randn(100, 4)
>>> arr[arr < .8] = pd.np.nan
>>> pd.DataFrame(arr).ftypes
0    float64:dense
1    float64:dense
2    float64:dense
3    float64:dense
dtype: object
>>> pd.SparseDataFrame(arr).ftypes
0    float64:sparse
1    float64:sparse
2    float64:sparse
3    float64:sparse
dtype: object

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
	No summary found (a short summary in a single line should be present at the beginning of the docstring)
	No returns section found
	Missing description for See Also "dtypes" reference
	Missing description for See Also "SparseDataFrame" reference

```

If the validation script still gives errors, but you think there is a good reason
to deviate in this case (and there are certainly such cases), please state this
explicitly.

Lastly, I left errors already occurred in the previous version without changes.

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -4275,7 +4275,29 @@ def get_ftype_counts(self):
 
     @property
     def dtypes(self):
-        """Return the dtypes in this object."""
+        """
+        Return the dtypes in this object.
+
+        Notes
```
Comment (TomAugspurger) [inline pandas/core/generic.py]: I think you can remove the "Notes" header, and just make this the extended summary.

File: pandas/core/generic.py [hunk 2]
```diff
@@ -4275,7 +4275,29 @@ def get_ftype_counts(self):
...
+        -----
+        It returns a Series with the data type of each column.
```
Comment (TomAugspurger) [inline pandas/core/generic.py]: Maybe replace "it" with "This method". And let's say what the values and index is.

    This returns a Series with the data type of each column. The result's index is the
    original DataFrame's columns.

Let's also replace all instances of "object" with "DataFrame". I'm not sure why this is in generic.py since I think it's specific to DataFrame.

File: pandas/core/generic.py [hunk 3]
```diff
@@ -4275,7 +4275,29 @@ def get_ftype_counts(self):
...
+        If object contains data multiple dtypes in a single column,
+        dtypes will be chosen to accommodate all of the data types.
+        ``object`` is the most general.
+
+        Examples
+        --------
+        >>> df = pd.DataFrame({'f': pd.np.random.rand(3),
```
Comment (TomAugspurger) [inline pandas/core/generic.py]: Can you just write out three floating point values? I'd like to avoid random data.

And FYI in general you don't want to use `pd.np`. You'll want to import NumPy manually (it's assumed to be imported in our docstrings.)

Comment (TomAugspurger) [reply pandas/core/generic.py]: Ah, and could you just make all these length-1 lists, just to be clearer? so `{'f': [1.0], 'i': [1], ...}`

File: pandas/core/generic.py [hunk 4]
```diff
@@ -4285,6 +4307,31 @@ def ftypes(self):
         """
         Return the ftypes (indication of sparse/dense and dtype)
         in this object.
+
+        Notes
+        -----
+        Sparse data should have the same dtypes as its dense representation
```
Comment (TomAugspurger) [inline pandas/core/generic.py]: End in a `.`

File: pandas/core/generic.py [hunk 5]
```diff
@@ -4285,6 +4307,31 @@ def ftypes(self):
...
+        --------
+        >>> arr = pd.np.random.randn(100, 4)
+        >>> arr[arr < .8] = pd.np.nan
+        >>> pd.DataFrame(arr).ftypes
+        0    float64:dense
+        1    float64:dense
+        2    float64:dense
+        3    float64:dense
+        dtype: object
+        >>> pd.SparseDataFrame(arr).ftypes
```
Comment (TomAugspurger) [inline pandas/core/generic.py]: Maybe a blank line before this to break things up a bit.

File: pandas/core/generic.py [hunk 6]
```diff
@@ -4285,6 +4307,31 @@ def ftypes(self):
...
+        -----
+        Sparse data should have the same dtypes as its dense representation
+
+        See Also
+        --------
+        dtypes, SparseDataFrame
+
+        Examples
+        --------
+        >>> arr = pd.np.random.randn(100, 4)
```
Comment (TomAugspurger) [inline pandas/core/generic.py]: `arr = np.random.RandomState(0).randn(100, 4)` for reproducibility

File: pandas/core/generic.py [hunk 7]
```diff
@@ -4275,7 +4275,29 @@ def get_ftype_counts(self):
 
     @property
     def dtypes(self):
-        """Return the dtypes in this object."""
+        """
+        Return the dtypes in this object.
+
```
Comment (DataOmbudsman) [inline pandas/core/generic.py]: Can you please add a Returns section as specified in the guide?

File: pandas/core/generic.py [hunk 8]
```diff
@@ -4275,7 +4275,29 @@ def get_ftype_counts(self):
...
+        Notes
+        -----
+        It returns a Series with the data type of each column.
+        If object contains data multiple dtypes in a single column,
+        dtypes will be chosen to accommodate all of the data types.
+        ``object`` is the most general.
+
```
Comment (DataOmbudsman) [inline pandas/core/generic.py]: I believe in a See Also section `ftypes` could be mentioned.

File: pandas/core/generic.py (hunk 2 above)
Comment (joaoavf) [inline pandas/core/generic.py]: Create a 'Returns' section to explain the output. Notes should be used only to explain technical details about the implementation of the algorithm or function behavior.

File: pandas/core/generic.py [hunk 9]
```diff
@@ -4285,6 +4307,31 @@ def ftypes(self):
         """
         Return the ftypes (indication of sparse/dense and dtype)
```
Comment (DataOmbudsman) [inline pandas/core/generic.py]: The docstring guide asks that the summary should fit in a single line. Could you rephrase it that way? If all information  could not fit in a single line you can then use an Extended Summary section.

File: pandas/core/generic.py [hunk 10]
```diff
@@ -4275,7 +4275,29 @@ def get_ftype_counts(self):
 
     @property
     def dtypes(self):
-        """Return the dtypes in this object."""
+        """
+        Return the dtypes in this object.
```
Comment (joaoavf) [inline pandas/core/generic.py]: Explain in more depth to a novice user that this is used to get the dtypes per column of the DataFrame.

File: pandas/core/generic.py [hunk 11]
```diff
@@ -4275,7 +4275,29 @@ def get_ftype_counts(self):
...
+
+        Notes
+        -----
+        It returns a Series with the data type of each column.
+        If object contains data multiple dtypes in a single column,
+        dtypes will be chosen to accommodate all of the data types.
+        ``object`` is the most general.
```
Comment (joaoavf) [inline pandas/core/generic.py]: It is worth explaining that `str` will be represented as `object`.

File: pandas/core/generic.py [hunk 12]
```diff
@@ -4285,6 +4307,31 @@ def ftypes(self):
...
         Return the ftypes (indication of sparse/dense and dtype)
         in this object.
+
+        Notes
+        -----
+        Sparse data should have the same dtypes as its dense representation
+
+        See Also
+        --------
+        dtypes, SparseDataFrame
```
Comment (DataOmbudsman) [inline pandas/core/generic.py]: Instead of simply `dtypes` you should use `pandas.DataFrame.dtypes`. You can even try linking to dtypes with the notation found in the guide: :meth:\`pandas.Series.sum\`

**Review by DataOmbudsman — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 7 above)
Comment (joaoavf) [inline pandas/core/generic.py]: Add a 'See Also' section to contemplate common dtypes.

File: pandas/core/generic.py [hunk 13]
```diff
@@ -4285,6 +4307,31 @@ def ftypes(self):
         """
```
Comment (joaoavf) [inline pandas/core/generic.py]: It would better organized it there was pull request for dtypes and another pull request for ftypes.

File: pandas/core/generic.py (hunk 9 above)
Comment (joaoavf) [inline pandas/core/generic.py]: Short summary should have 1 line.

File: pandas/core/generic.py [hunk 14]
```diff
@@ -4285,6 +4307,31 @@ def ftypes(self):
...
         Return the ftypes (indication of sparse/dense and dtype)
         in this object.
+
```
Comment (joaoavf) [inline pandas/core/generic.py]: Add a return section.

File: pandas/core/generic.py [hunk 15]
```diff
@@ -4285,6 +4307,31 @@ def ftypes(self):
...
+        Notes
+        -----
+        Sparse data should have the same dtypes as its dense representation
+
+        See Also
```
Comment (joaoavf) [inline pandas/core/generic.py]: 'See Also' should go before notes.

File: pandas/core/generic.py (hunk 12 above)
Comment (joaoavf) [inline pandas/core/generic.py]: Make a short description for the itens cited in the summary.

**Review by joaoavf — COMMENTED** [review]
Found some points that could be improved. Good sprint for you all!

**Comment by jreback** [comment]
is this overlapping with #20099 ?

File: pandas/core/generic.py [hunk 16]
```diff
@@ -4275,7 +4275,29 @@ def get_ftype_counts(self):
...
+        If object contains data multiple dtypes in a single column,
+        dtypes will be chosen to accommodate all of the data types.
+        ``object`` is the most general.
+
+        Examples
+        --------
+        >>> df = pd.DataFrame({'f': pd.np.random.rand(3),
+        ...                    'i': 1,
+        ...                    'd': pd.Timestamp('20180310'),
+        ...                    'o': 'foo'})
```
Comment (jreback) [inline pandas/core/generic.py]: add in See Also to Series.dtype

File: pandas/core/generic.py (hunk 2 above)
Comment (lastone9182) [reply pandas/core/generic.py]: I'm modifying dtype `@property` of the NDFrame in generic.py
Is "This property" okay instead of "This method"?

At first, ``object`` was ``NDFrame``, but found an error.

```
Private classes (['NDFrame']) should not be mentioned in public docstring.
```

but it seems to be better that the way you specify with "DataFrame", I will do that.
